lib/ddtrace/profiling/collectors/stack.rb

```diff
@@ -0,0 +1,111 @@
+require 'ddtrace/profiling/events/stack'
+require 'ddtrace/utils/time'
+require 'ddtrace/worker'
+require 'ddtrace/workers/polling'
+
+module Datadog
+  module Profiling
+    module Collectors
+      # Pulls and exports profiling data on an async interval basis
+      class Stack < Worker
+        include Workers::Polling
+
+        MIN_INTERVAL = 0.01
+        MAX_FRAMES = 128
+
+        attr_reader \
+          :ignore_thread,
+          :last_wall_time,
+          :max_frames,
+          :max_time_usage_pct,
+          :recorder
+
+        def initialize(recorder, options = {})
+          @recorder = recorder
+          @max_frames = options.fetch(:max_frames, MAX_FRAMES)
+          @ignore_thread = options.fetch(:ignore_thread, nil)
+          @max_time_usage_pct = options.fetch(:max_time_usage_pct, 2.0)
+
+          # Workers::Async::Thread settings
+          # Restart in forks by default
+          self.fork_policy = options.fetch(:fork_policy, Workers::Async::Thread::FORK_POLICY_RESTART)
+
+          # Workers::IntervalLoop settings
+          self.loop_base_interval = options.fetch(:interval, MIN_INTERVAL)
+
+          # Workers::Polling settings
+          self.enabled = options.fetch(:enabled, false)
+        end
+
+        def start
+          perform
+        end
+
+        def perform
+          collect_and_wait
+        end
+
+        def loop_back_off?
+          false
+        end
+
+        def collect_and_wait
+          run_time = Datadog::Utils::Time.measure do
+            collect_events
+          end
+
+          # Update wait time to throttle profiling
+          self.loop_wait_time = compute_wait_time(run_time)
+        end
+
+        def collect_events
+          @last_wall_time ||= Datadog::Utils::Time.get_time
```

Comment:
This is kind of awkward for the first loop, but we don't want the wall time to be `nil` or the math below will raise an error.

Reply:
Would it make sense to initialize `@last_wall_time` at `#start`, for example, instead?

Reply:
I don't think so, because `start` is just an alias for `perform`, and technically this collection can be run synchronously with `collect_events`. If this happens, `@last_wall_time` will be `nil` and it will raise an error. Better to put it closer to where the code actually uses it I think.

Reply:
Ok, if we don't have a clean starting point to track the initial `@last_wall_time`, we would probably then throw away the first measurement of `collect_events`, likely by returning an empty `events = []` early.

I'm thinking we don't want a stack track recorded that ran in almost zero `wall_time_interval_ns`, as we might later have to divide some measurement by time (`wall_time_interval_ns` time) and get an exorbitant result because of this.

I might be wrong here, and `wall_time_interval_ns` might not cause such issues when mis-measured for the first list of stack traces. I think it's something @jd might know more about.

Reply:
You could actually do both I think: if when started you don't collect first but start by sleeping, you can set `@last_wall_time` at that point.

For the Python collector, I just dismissed the "sync" case where you don't collect/sleep. If I had to support it I'd probably do what @delner did there — the time barely makes sense because it's very small, but the fact that you would do a single discrete collect call is also weird. :)

Reply:
Yeah, maybe it doesn't make sense to do a "sync" call... this might be leftover thinking from when I was working on the scheduler (aka flush before shutdown) that might not apply here at all.

Nonetheless, there are multiple paths to `collect_events`, and `start` is only one of them, which could result in an exception and crash the worker if time isn't initialized.

As a compromise, maybe we have a lazy initialization of last wall time as a safety measure, and have `start` set it to its own value. As I see it the first measurement is probably going to always be garbage anyways (as there's no true point of reference), and this can run as fast as every 10ms, so I don't see any inaccuracy as very significant or meaningful.